tests/ui_tools/test_boxes.py

```diff
@@ -743,74 +743,105 @@ def test_generic_autocomplete_user_mentions(
+            ("#S", 0, "#**Secret stream**", {}),  # 1st-word startswith match.
+            ("#S", 1, "#**Some general stream**", {}),  # 1st-word startswith.
+            ("#S", 2, "#**Stream 1**", {}),  # 1st-word startswith match.
+            ("#S", 3, "#**Stream 2**", {}),  # 1st-word startswith match.
+            ("#S", 4, "#**Web public stream**", {}),  # 1st-word startswith match.
+            ("#S", -1, "#**Web public stream**", {}),
+            ("#S", -2, "#**Stream 2**", {}),
+            ("#S", -3, "#**Stream 1**", {}),
+            ("#S", -4, "#**Some general stream**", {}),
+            ("#S", -5, "#**Secret stream**", {}),
+            ("#S", -6, None, {}),
+            ("#So", 0, "#**Some general stream**", {}),
+            ("#So", 1, None, {}),
+            ("#Se", 0, "#**Secret stream**", {}),
+            ("#Se", 1, None, {}),
+            ("#St", 0, "#**Stream 1**", {}),
+            ("#St", 1, "#**Stream 2**", {}),
+            ("#g", 0, "#**Some general stream**", {}),
+            ("#g", 1, None, {}),
+            ("#Stream 1", 0, "#**Stream 1**", {}),  # Complete match.
+            ("#nomatch", 0, None, {}),
+            ("#ene", 0, None, {}),
             # Complex autocomplete prefixes.
-            ("[#Stream", 0, "[#**Stream 1**", []),
-            ("(#Stream", 1, "(#**Stream 2**", []),
-            ("@#Stream", 0, "@#**Stream 1**", []),
-            ("@_#Stream", 0, "@_#**Stream 1**", []),
-            (":#Stream", 0, ":#**Stream 1**", []),
-            ("##Stream", 0, "##**Stream 1**", []),
-            ("##*Stream", 0, None, []),  # NOTE: Optional single star fails
-            ("##**Stream", 0, "##**Stream 1**", []),  # Optional 2-stars
+            ("[#Stream", 0, "[#**Stream 1**", {}),
+            ("(#Stream", 1, "(#**Stream 2**", {}),
+            ("@#Stream", 0, "@#**Stream 1**", {}),
+            ("@_#Stream", 0, "@_#**Stream 1**", {}),
+            (":#Stream", 0, ":#**Stream 1**", {}),
+            ("##Stream", 0, "##**Stream 1**", {}),
+            ("##*Stream", 0, None, {}),  # NOTE: Optional single star fails
+            ("##**Stream", 0, "##**Stream 1**", {}),  # Optional 2-stars
             # With 'Secret stream' pinned.
             (
                 "#Stream",
                 0,
                 "#**Secret stream**",
-                ["Secret stream"],
+                {"pinned": ["Secret stream"]},
             ),  # 2nd-word startswith match (pinned).
             (
                 "#Stream",
                 1,
                 "#**Stream 1**",
-                ["Secret stream"],
+                {"pinned": ["Secret stream"]},
             ),  # 1st-word startswith match (unpinned).
             (
                 "#Stream",
                 2,
                 "#**Stream 2**",
-                ["Secret stream"],
+                {"pinned": ["Secret stream"]},
             ),  # 1st-word startswith match (unpinned).
```

Comment:
I think it may be clearer if we remove these comments in a separate commit (maybe when adding the web public cases?).

However, see my general review point too.